### ISSUE TITLE
Type checking worked before, now doesn't for generic functions that take return type from a runtime parameter

### ISSUE THREAD

**Issue by zackyancey** [microsoft/pyright#7569]
**Describe the bug**

I have a function that takes as one of its arguments a type which is used at runtime to determine the return type of the function (specifically, the type argument is passed to pydantic to create a parser). This used to typecheck properly, but now it doesn't if the type is a union:

**Code or Screenshots**

```python
from typing import TypeVar

T = TypeVar("T")

def parse(into_type: type[T], json: str) -> T:
    ...

# This one works, no errors and `a` gets the `int` type.
a = parse(int, "23")

# Previously: also no errors, `b` is `int | None`
# Now, Error: Argument of type "type[int] | type[None]" cannot be assigned to parameter 
# "into_type" of type "type[T@parse]" in function "parse"
#    Type "UnionType" cannot be assigned to type "type[T@parse]"
b = parse(int | None, "null")
```

**VS Code extension or command-line**

Observed in both. Narrowed it down with the CLI, it worked in 1.1.345 and stopped working in 1.1.346

**Comment by erictraut** [comment]
The new behavior is correct, so this isn't a bug. The older behavior was not compliant with the typing spec. A union is not compatible with `type[T]` because it is not an instance of `type` and is not callable.

There is discussion within the typing forums to add a new special form called `TypeForm` that would allow you to handle your use case, but the details still need to be worked out before a PEP can be written and implemented.
